Contributing (#4133)
mixxxdj/mixxx: CONTRIBUTING.md

# Mixxx Contribution Guidelines #

> Thank you for contributing to [Mixxx](https://mixxx.org/)! Your work helps DJs all over the world! We are a global team that works by consensus and we are excited to have you join us.

Reviewer: @Be-ing What do you think?

> Thank you for contributing to [Mixxx](https://mixxx.org/)! Your work helps DJs all over the world! We are a global team that works by consensus and we are excited to have you join us.

Reviewer 2: :shrug:

> All Mixxx contributors, including the core team members, are volunteers. This means that Mixxx is made possible by their passion and commitment to the project on their free time. This is why it's so important to treat everyone with respect and understanding, and be patient if things sometimes move a bit slower than you would like. Don't forget to read and follow our [Code Of Conduct](https://github.com/mixxxdj/mixxx/blob/master/CODE_OF_CONDUCT.md), as it provides a framework for everyone to feel accepted, respected and safe in all their interactions within the mixxx community.

Reviewer 2: ```suggestion
All Mixxx contributors, including the core team members, are volunteers. This means that Mixxx is made possible by their passion and commitment to the project on their free time. This is why it's so important to treat everyone with respect and understanding, and be patient if things sometimes move a bit slower than you would like. Please read and follow our [Code Of Conduct](https://github.com/mixxxdj/mixxx/blob/master/CODE_OF_CONDUCT.md), as it provides a framework for everyone to feel accepted, respected and safe in all their interactions within the Mixxx community.
```

## Table of Contents ##
- [Getting in touch with the community](#getting-in-touch-with-the-community)
- [How can I contribute?](#how-can-i-contribute)
- [Your first Mixxx code contribution](#Your-first-Mixxx-code-contribution)
   - [Compile Mixxx for the first time](#Compile-Mixxx-for-the-first-time)
   - [Pick an issue to work on](#Pick-an-issue-to-work-on)
   - [Time to code](#Time-to-code)
   - [Open a Pull Request](#Open-a-Pull-Request)
   - [Contributor Agreement](#Contributor-Agreement)
   - [How to become a 5-star contributor](#How-to-become-a-5-star-contributor)

## Getting in touch with the community ##

We use [Zulip chat](https://mixxx.zulipchat.com/) to discuss and coordinate everything related to Mixxx development (and more!).

> - We encourage you to say hi on the [introduce yourself channel](https://mixxx.zulipchat.com/#narrow/stream/109123-introduce-yourself) and share what you would like to contribute.

Reviewer 2: ```suggestion
- We encourage you to say hi on the [introduce yourself stream](https://mixxx.zulipchat.com/#narrow/stream/109123-introduce-yourself) and share what you would like to contribute.
```

> - If you need help with your contribution, the Mixxx codebase or anything else related to Mixxx development don't hesitate to ask on the [development help channel](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).

Reviewer 2: ```suggestion
- If you need help with your contribution, the Mixxx codebase or anything else related to Mixxx development don't hesitate to ask on the [development help stream](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
```

> - The [development channel](https://mixxx.zulipchat.com/#narrow/stream/109171-development) is where day to day discussion about code, new features and releases happens.

Reviewer 2: ```suggestion
- The [development stream](https://mixxx.zulipchat.com/#narrow/stream/109171-development) is where day to day discussion about code, new features and releases happens.
```

We also have lots of additional information for users and developers on the [Mixxx wiki](https://github.com/mixxxdj/mixxx/wiki), make sure to check it out if there's some piece of information you can't find.

## How can I contribute? ##

There are lots of ways you can contribute to Mixxx. Even if you don't know how to program, there's plenty of things you can do to help:

 - [Report a bug](https://github.com/mixxxdj/mixxx/wiki/Reporting%20bugs)

>  - [Help testing the latest developments](https://github.com/mixxxdj/mixxx/wiki/Testing)

Reviewer 2: ```suggestion
 - [Test the latest developments](https://github.com/mixxxdj/mixxx/wiki/Testing)
```

>  - [Help translate Mixxx to other languages](https://github.com/mixxxdj/mixxx/wiki/Internationalization)

Reviewer 2: ```suggestion
 - [Translate Mixxx to other languages](https://github.com/mixxxdj/mixxx/wiki/Internationalization)
```

>  - [Improve the Mixxx Manual](https://github.com/mixxxdj/manual#readme)

Reviewer 2: ```suggestion
 - [Improve the Mixxx manual](https://github.com/mixxxdj/manual#readme)
```

 - [Improve the Mixxx website](https://github.com/mixxxdj/website)

If you have some coding skills, or you are willing to learn them you can do even more:

>  - [Add native support for new hardware controllers](https://github.com/mixxxdj/mixxx/wiki/Contributing%20mappings)

Reviewer 2: ```suggestion
 - [Add a mapping for a new controller](https://github.com/mixxxdj/mixxx/wiki/Contributing%20mappings)
```

 - [Fix bugs or program new features](#Your-first-Mixxx-code-contribution)

We welcome programmers of all backgrounds and skill levels, and our community will do their best to help you with any problems you encounter.

## Your first Mixxx code contribution ##

> ### Compile Mixxx for the first time ###

Reviewer 3: We need to make sure that contributors create a GitHub account + a Mixxx fork first. Than clone the private repro.
Following the links here doe not emphasis this.
It is the most common issue in contributions from not experienced GitHub users.

Reviewer: I updated the linked wiki pages explaining this. 

We can go further and just instruct people to fork, not to clone our repo. What do you think?

Reviewer 3: Yes, good idea

Reviewer: Done

Your first goal is to get the Mixxx source code and compile it. The steps to do so depend on your operating system:
- [Compiling on Linux](https://github.com/mixxxdj/mixxx/wiki/Compiling%20on%20Linux)
- [Compiling on Windows](https://github.com/mixxxdj/mixxx/wiki/Compiling%20on%20Windows)
- [Compiling on macOS](https://github.com/mixxxdj/mixxx/wiki/Compiling%20on%20macOS)

### Pick an issue to work on ###
You might already have in mind that feature you always dreamed of! If that's not the case or you want to start with something easy, you can ask for ideas on [Zulip](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help) or pick an [easy issue](https://bugs.launchpad.net/mixxx/+bugs?field.tag=easy).

### Time to code ###
Now it's time to write some code!
We recommend to [set up an IDE](https://github.com/mixxxdj/mixxx/wiki/Developer%20Tools) to assit you in the code writing process.

> We use [Git](https://git-scm.com/) for source control. Git is a distributed version control system that allows us to keep track of how our code changes and work together on it. If you are new to Git or source control in general, you can have a look at [Using Git](https://github.com/mixxxdj/mixxx/wiki/Using%20Git) to learn how.

Reviewer 3: This can be moved up an shorten a bit.
The user most likely reads this file on GitHub.

> If you have trouble with any of this remember that you can always ask on the [Zulip development help channel](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).

Reviewer 2: ```suggestion
If you have trouble with any of this remember that you can always ask on the [Zulip development help stream](https://mixxx.zulipchat.com/#narrow/stream/247620-development-help).
```

### Open a Pull Request ###

> Once you finish coding, it's time to show your changes to other Mixxx developers. You do so by [opening a Pull Request](https://github.com/mixxxdj/mixxx/wiki/Using%20Git#open-a-pull-request).

Reviewer 2: ```suggestion
Once you finish coding, it's time to show your changes to other Mixxx developers. You do so by [opening a pull request](https://github.com/mixxxdj/mixxx/wiki/Using%20Git#open-a-pull-request).
```

### Contributor Agreement ###

> Because Mixxx is distributed through the macOS AppStore, in order to distribute your changes with Mixxx, we need a record of your permission to distribute them under Mixxx's license. You'll have to read and sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ).

Reviewer 2: ```suggestion
In the past Mixxx was distributed in the Mac App Store, which has terms conflicting with Mixxx's license, the GNU GPL (version 2 or later). In case we decide to distribute Mixxx in Apple's App Store again, in order to distribute your changes with Mixxx, we need a record of your permission to distribute them under Mixxx's license. You'll have to read and sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ).
```

* Controller mappings should be targeted at the stable branch unless they use features that are new in the beta or master branch.
* If you are making changes to the GUI with a pull request, please post before and after screenshots of the changes.
* Please help review other people's pull requests. When others review your pull requests, please return the favor. The continued progress of Mixxx depends on all of us working together. Even if you are not familiar with the area of the code being changed in a pull request, you can be helpful by building the branch, verifying that it works as described, and commenting with feedback about the user experience design.
* If you demonstrate good coding skills, help review pull requests, contribute major features, and show a commitment to Mixxx over time, we may invite you to the core team.

### Core Team ###
#### Core Team ####
Mixxx core team members are contributors who have write access to the [upstream mixxxdj repositories](https://github.com/mixxxdj/) on GitHub, access to the Jenkins web interface for the build servers, and access to the private Zulip stream for the core team.

> * Enable [two-factor authentication (2FA)](https://help.github.com/en/github/authenticating-to-github/securing-your-account-with-two-factor-authentication-2fa) for your GitHub account.

Reviewer 2: Why did you remove this information?

Reviewer: I move to the linked wiki page. This is not something the majority of users need to worry about.

Reviewer 2: I'd prefer to move documentation into the code repository or manual from the wiki.

Reviewer 4: I agree with this change, CONTRIBUTING should only be an introduction, and that section does not belong into the manual either
